File: src_cpp/grid.cpp
#include "grid.h"

Grid::Grid(int inp_M, int inp_k) {
    M = inp_M;
    k = inp_k;

    ancillas = new mat<int>(M, M, k);
    dests = new mat<bool>(M, M, false);
    in_progress = new mat<int>(M, M, 0);

    vector<Generator> generators;
    vector<Generator> finished_gens;
    vector<vector<vector<Point>>> full_chains;
    vector<vector<tuple<Point, Point>>> bell_pairs;
}

Grid::~Grid() {
    delete ancillas;
    delete dests;
    delete in_progress;
}

const mat<int>& Grid::get_ancillas() {
    return *ancillas;
}

vector<Point> Grid::find_chain(const mat<int>& grid, Point site1, Point site2) {
    mat<bool> visited(M, M, false);
    mat<Point> parent(M, M, Point());
    deque<Point> queue = {site1};
    visited(site1.x, site1.y) = true;

    if ((grid(site1.x, site1.y) <= 0) || 
        (grid(site2.x, site2.y) <= 0)) {
        return {};
    }

    while (queue.size()) {
        Point curr_site = queue.front();
        int x = curr_site.x, y = curr_site.y;
        queue.pop_front();

        if (curr_site == site2) {
            vector<Point> chain = {curr_site};
            while (curr_site != site1) {
                curr_site = parent(curr_site.x, curr_site.y);
                chain.insert(begin(chain), curr_site);
            }
            return chain;
        }

        if ((curr_site != site1) && (grid(x, y) < 2)) { 
            continue; 
        }
        vector<Point> pot_nbrs = { Point(x, y+1), Point(x, y-1), Point(x+1, y), Point(x-1, y) };
        for (auto it = begin(pot_nbrs); it != end(pot_nbrs); it++) {
            int new_x = (*it).x, new_y = (*it).y;

            if ((0 <= new_x && new_x < M) && (0 <= new_y && new_y < M)) {
                if (!visited(new_x, new_y)) {
                    parent(new_x, new_y) = curr_site;
                    queue.push_back(*it);
                    visited(new_x, new_y) = true;
                }
            }
        }

    }

    return {};
}

void Grid::add_chain(vector<Point> chain, int gen) {
    for (size_t i = 0; i < chain.size(); i++) {
        int x = chain[i].x, y = chain[i].y;
        if ((i == 0) || (i == (chain.size() - 1))) {
            (*ancillas)(x, y) = (*ancillas)(x, y) - 1;
        } else {
            (*ancillas)(x, y) = (*ancillas)(x, y) - 2;
        }
    }
    full_chains[gen].push_back(chain);
}

void Grid::tmp_add_chain(mat<int>& grid, vector<Point> chain) {
    for (size_t i = 0; i < chain.size(); i++) {
        int x = chain[i].x, y = chain[i].y;
        if ((i == 0) || (i == (chain.size() - 1))) {
            grid(x, y) = grid(x, y) - 1;
        } else {
            grid(x, y) = grid(x, y) - 2;
        }
    }
}

void Grid::perform_bell_measurement() {
    for (size_t i = 0; i < generators.size(); i++) {
        vector<vector<Point>> gen_chains = full_chains[generators[i].key];
        for (size_t j = 0; j < gen_chains.size(); j++) {
            vector<Point> chain = gen_chains[j];
            for (size_t k = 1; k < (chain.size() - 1); k++) {
                int x = chain[k].x, y = chain[k].y;
                (*ancillas)(x, y) = (*ancillas)(x, y) + 2;
            }
            bell_pairs[generators[i].key].push_back(make_tuple(chain[0], chain[chain.size()-1]));
        } 
        full_chains[generators[i].key] = vector<vector<Point>>();
    }
}

void Grid::perform_syndrome_measurements() {
    for (size_t i = 0; i < generators.size(); i++) {
        Generator gen = generators[i];
        if (gen.is_done()) {
            int x = gen.dest.x, y = gen.dest.y;
            (*ancillas)(x, y) = (*ancillas)(x, y) + 1;
            (*dests)(x, y) = false;

            vector<tuple<Point, Point>> pairs = bell_pairs[gen.key];
            for (size_t j = 0; j < pairs.size(); j++) {
                const auto[qbt1, qbt2] = pairs[j];
                int x1 = qbt1.x, y1 = qbt1.y;
                (*ancillas)(x1, y1) = (*ancillas)(x1, y1) + 1;
                int x2 = qbt2.x, y2 = qbt2.y;
                (*ancillas)(x2, y2) = (*ancillas)(x2, y2) + 1;
            }

            bell_pairs[gen.key] = vector<tuple<Point, Point>>();

            vector<Point> qbts = gen.all_qbts;
            for (size_t j = 0; j < qbts.size(); j++) {
                Point qbt = qbts[j];
                (*in_progress)(qbt.x, qbt.y) = (*in_progress)(qbt.x, qbt.y) - 1;
            }
        }
    }
}

void Grid::reset_generator(Generator& gen) {
    if ((!gen.is_done()) && (gen.dest != Point())) { // not done but started routing
        int x = gen.dest.x, y = gen.dest.y;
        (*ancillas)(x, y) = (*ancillas)(x, y) + 1;
        (*dests)(x, y) = false;

        vector<tuple<Point, Point>> pairs = bell_pairs[gen.key];
        for (size_t j = 0; j < pairs.size(); j++) {
            const auto[qbt1, qbt2] = pairs[j];
            int x1 = qbt1.x, y1 = qbt1.y;
            (*ancillas)(x1, y1) = (*ancillas)(x1, y1) + 1;
            int x2 = qbt2.x, y2 = qbt2.y;
            (*ancillas)(x2, y2) = (*ancillas)(x2, y2) + 1;
        }

        bell_pairs[gen.key] = vector<tuple<Point, Point>>();

        vector<Point> qbts = gen.all_qbts;
        for (size_t j = 0; j < qbts.size(); j++) {
            Point qbt = qbts[j];
            (*in_progress)(qbt.x, qbt.y) = (*in_progress)(qbt.x, qbt.y) - 1;
        }
    }

    gen.reset();
}

vector<vector<Point>> Grid::route_generator(vector<Point> gen, Point prior_dest) {
    bool is_prior_dest = (prior_dest != Point());
    vector<vector<vector<Point>>> out = {};
    vector<Point> possible_dests = gen;
    if (is_prior_dest) {
        possible_dests = {prior_dest};
    }

    for (size_t i = 0; i < possible_dests.size(); i++) {
        Point dest = possible_dests[i];
        int x = dest.x, y = dest.y;

        // prevent gridlock, could have parameter turning this off
        if (!is_prior_dest && ((*in_progress)(x, y))) continue;

        mat<int> tmp_grid = get_ancillas();
        vector<vector<Point>> chains = {};
        vector<Point> routed_qbts = {};
        size_t tot_len = 0;

        // dest is the meeting site
        if (!is_prior_dest) {
            tmp_grid(x, y) = tmp_grid(x, y) - 1;
        }

        for (size_t j = 0; j < gen.size(); j++) {
            Point site = gen[j];
            
            if ((dest != site) && (!(*dests)(site.x, site.y))) {
                vector<Point> chain = find_chain(tmp_grid, dest, site);
                if (chain.size()) {
                    tmp_add_chain(tmp_grid, chain);
                    chains.push_back(chain);
                    routed_qbts.push_back(site);
                    tot_len = tot_len + chain.size();
                } else continue;
            }
        }
        out.push_back(chains);
    }

     if (!is_prior_dest) {
        stable_sort(begin(out), end(out), []
            (const vector<vector<Point>>& lhs, const vector<vector<Point>>& rhs) {
                if (lhs.size() < rhs.size()) return false;
                if (rhs.size() < lhs.size()) return true;

                size_t tot_lhs_size = 0;
                size_t tot_rhs_size = 0;
                for (size_t i = 0; i < lhs.size(); i++) {
                    tot_lhs_size += lhs[i].size();
                }
                for (size_t i = 0; i < rhs.size(); i++) {
                    tot_rhs_size += rhs[i].size();
                }
                if (tot_lhs_size < tot_rhs_size) return true;
                if (tot_rhs_size < tot_lhs_size) return false;
                return false;
            });
    }

    // for (size_t i = 0; i < out.size(); i++) {
    //     vector<vector<Point>> chains = out[i];
    //     for (size_t j = 0; j < chains.size(); j++) {
    //         vector<Point> chain = chains[j];
    //         for (auto it = begin(chain); it != end(chain); it++) {
    //             cout << "(" << (*it).x << " " << (*it).y << ")";
    //         }
    //         cout << endl;
    //     }
    //     cout << "........." << endl;
    // }

    if (out.size()) {
        return out[0];
    } else {
        return {};
    }
}

int Grid::greedy_route_set(vector<vector<Point>> gens) {
    for (size_t i = 0; i < gens.size(); i++) {
        if (int(gens[i].size()) > k) return 0;

        full_chains.push_back(vector<vector<Point>>());
        bell_pairs.push_back(vector<tuple<Point, Point>>());

        Generator gen(gens[i], i);
        generators.push_back(gen);
    }

    // stable_sort(begin(generators), end(generators), []
    //     (const Generator& lhs, const Generator& rhs) {
    //         return lhs.c.r < rhs.c.r;
    //     });

    for (size_t rounds = 0; rounds < (gens.size() + 1); rounds++) {
        stable_sort(begin(generators), end(generators), []
            (const Generator& lhs, const Generator& rhs) {
                if (lhs.num_routed() == rhs.num_routed()) {
                    // return lhs.c.r < rhs.c.r;
                    return false;
                } else {
                    return lhs.num_routed() > rhs.num_routed();
                }
            });

        for (size_t i = 0; i < generators.size(); i++) {
            Generator &gen = generators[i];
            vector<vector<Point>> chains = route_generator(gen.qbts_to_route, gen.dest);

            if (chains.size()) {
                if (gen.dest == Point()) {
                    Point dest = chains[0][0];
                    int x = dest.x, y = dest.y;
                    gen.dest = dest;
                    (*ancillas)(x, y) = (*ancillas)(x, y) - 1;
                    (*dests)(x, y) = true;
                    gen.route_qbt(dest);
                    gen.start = rounds;

                    vector<Point> qbts = gen.all_qbts;
                    for (size_t j = 0; j < qbts.size(); j++) {
                        Point qbt = qbts[j];
                        (*in_progress)(qbt.x, qbt.y) = (*in_progress)(qbt.x, qbt.y) + 1;
                    }
                }

                for (size_t j = 0; j < chains.size(); j++) {
                    vector<Point> chain = chains[j];
                    gen.route_qbt(chain[chain.size() - 1]);
                    add_chain(chain, gen.key);
                }
            }      
            // reset_generator(gen);

        }

        // double sum = 0;
        // for (int i = 0; i < M; i++) {
        //     for (int j = 0; j < M; j++) {
        //         sum += (*ancillas)(i, j);
        //     }
        // }
        // cout << rounds << "," << sum << "," << M*M*k << "," << sum/double(M*M*k) << endl;
          
        perform_bell_measurement();
        perform_syndrome_measurements();

        for (size_t i = 0; i < generators.size(); i++) {
            Generator gen = generators[i];
            if (gen.is_done()) {
                cout << gen.c.r << "," << rounds+1 << "," << rounds+1-gen.start << endl;   
            }
        }

        generators.erase(remove_if(begin(generators), end(generators), 
            [](const Generator& gen) {
                return gen.is_done();
            }), end(generators));
        if (!generators.size()) {
            return rounds+1;
        }        
    }

    return -1;
}


void Grid::greedy_route_reccuring(vector<vector<Point>> gens, int max_rounds, int restart) {
    for (size_t i = 0; i < gens.size(); i++) {
        if (int(gens[i].size()) > k) return;

        full_chains.push_back(vector<vector<Point>>());
        bell_pairs.push_back(vector<tuple<Point, Point>>());

        Generator gen(gens[i], i);
        generators.push_back(gen);
    }

    // what info do we want out of this function? 
    // every time a generator is restarted, print round, cycle, and if it finished...
    // maybe when adding back in you should shuffle the generators
    for (int rounds = 1; rounds < max_rounds; rounds++) {
        stable_sort(begin(generators), end(generators), []
            (const Generator& lhs, const Generator& rhs) {
                if (lhs.num_routed() == rhs.num_routed()) {
                    // return lhs.c.r < rhs.c.r;
                    return false;
                } else {
                    return lhs.num_routed() > rhs.num_routed();
                }
            });

        for (size_t i = 0; i < generators.size(); i++) {
            Generator &gen = generators[i];
            vector<vector<Point>> chains = route_generator(gen.qbts_to_route, gen.dest);

            if (chains.size()) {
                if (gen.dest == Point()) {
                    Point dest = chains[0][0];
                    int x = dest.x, y = dest.y;
                    gen.dest = dest;
                    (*ancillas)(x, y) = (*ancillas)(x, y) - 1;
                    (*dests)(x, y) = true;
                    gen.route_qbt(dest);

                    vector<Point> qbts = gen.all_qbts;
                    for (size_t j = 0; j < qbts.size(); j++) {
                        Point qbt = qbts[j];
                        (*in_progress)(qbt.x, qbt.y) = (*in_progress)(qbt.x, qbt.y) + 1;
                    }
                }

                for (size_t j = 0; j < chains.size(); j++) {
                    vector<Point> chain = chains[j];
                    gen.route_qbt(chain[chain.size() - 1]);
                    add_chain(chain, gen.key);
                }
            }      
        }

        

        perform_bell_measurement();
        perform_syndrome_measurements();

        double sum = 0;
        for (int i = 0; i < M; i++) {
            for (int j = 0; j < M; j++) {
                sum += (*ancillas)(i, j);
            }
        }


        for (auto it = begin(generators); it != end(generators);) {
            Generator &gen = *it;

            if (gen.is_done()) {
                finished_gens.push_back(gen);
                it = generators.erase(it);
            } else {
                it++;
            }
        }

        if ((rounds) && (rounds % restart == 0)) {
            cout << "ROUNDS:" << rounds << "," << sum << "," << M*M*k << "," << sum/double(M*M*k) << endl;

            for (auto it = begin(generators); it != end(generators); it++) {
                Generator &gen = *it;
                cout << gen.c.r << "," << gen.key << "," << rounds << "," << gen.cycle << "," << gen.is_done() << endl;
            }
            for (auto it = begin(finished_gens); it != end(finished_gens);) {
                Generator &gen = *it;

                cout << gen.c.r << "," << gen.key << "," << rounds << "," << gen.cycle << "," << gen.is_done() << endl;
                reset_generator(gen);

                generators.push_back(gen);
                it = finished_gens.erase(it);
            }
        }
    }
}


int Grid::route_independent_sets(vector<vector<Point>> gens) {
    // split the generators up into indepenent sets to be routed separately.
    // each independent set can be routed in one round. Total time is thus number of independent sets

    size_t orig_size = gens.size();
    for (size_t rounds = 0; rounds < (orig_size + 1); rounds++) {
        vector<vector<Point>> ind_set = {gens[0]};
        gens.erase(begin(gens));

        for (auto it = begin(gens); it != end(gens);) {
            bool add = true;

            for (auto it2 = begin(ind_set); it2 != end(ind_set); it2++) {
                vector<Point> intersection;
                set_intersection(begin(*it), end(*it),
                               begin(*it2), end(*it2),
                               back_inserter(intersection));

                if (intersection.size()) {
                    add = false;
                }
            }
            
            if (add) {
                ind_set.push_back(*it);
                it = gens.erase(it);
            } else {
                it++;
            }
        }

        if (!gens.size()) {
            return rounds+1;
        }
    }

    return -1;
}

File: src_py/Grid.py
from itertools import islice
import random
import numpy as np
import copy

class Generator:
    def __init__(self, qbts, dest, i):
        # have to route all qubits. if dest in generator, then it's essentially a free route
        self.qbts = qbts
        self.key = i
        self.qbts_to_route = qbts.copy()
        self.routed = []
        self.dest = dest # could be in generator or not


class Grid:
    # class that contains a grid of qubit (objects?), site objects perhaps containing data qubit and some number of ancillas
    # will be in charge of actual routing and generating sets to route

    def __init__(self, N, M, k):
        # construct a MxN grid of sites each with k ancilla qubits
        self.reset(N, M, k)

    def reset(self, N, M, k):
        self.N = N
        self.M = M
        self.k = k
        self.generators = []
        self.full_chains = {}
        self.bell_pairs = {}
        self.grid = [[k for _ in range(self.N)] for _ in range(self.M)]
        self.dests = [[False for _ in range(self.N)] for _ in range(self.M)]
        self.in_progress = [[0 for _ in range(self.N)] for _ in range(self.M)]


    def find_chain(self, grid, qSite1, qSite2):
        # this function checks to make sure that the intermediate sites have enough ancilla qubits
        # free to handle the Bell pair chain. Also checks endpoints of chain for one available ancilla
        # returns empty chain if impossible

        visited = [[False for x in range(self.N)] for y in range(self.M)]
        parent = [[None for x in range(self.N)] for y in range(self.M)]
        queue = [qSite1]
        visited[qSite1[0]][qSite1[1]] = True

        if ((grid[qSite1[0]][qSite1[1]] <= 0) or
            (grid[qSite2[0]][qSite2[1]] <= 0)):
            return []

        while queue:
            x, y = queue.pop(0)
            if ((x, y) == qSite2):

                curr_site = qSite2
                chain = [curr_site]

                while (not (curr_site == qSite1)):
                    curr_site = parent[curr_site[0]][curr_site[1]]
                    chain.append(curr_site)
                return chain[::-1]

            if (((x, y) != qSite1) and (grid[x][y] < 2)):
                continue
            pot_nbrs = [(x, y+1), (x, y-1), (x+1, y), (x-1, y)]
            for nbr in pot_nbrs:
                new_x, new_y = nbr
                if ((0 <= new_x < self.M) and (0 <= new_y < self.N)):
                    if (not visited[new_x][new_y]):
                        parent[new_x][new_y] = (x, y)
                        queue.append(nbr)
                        visited[new_x][new_y] = True

        return []


    def add_chain(self, chain, gen):
        # can assume that the path parameter is valid, i.e. there are enough ancillas available
        for i, pair in enumerate(chain):
            x, y = pair
            if ((i == 0) or (i == len(chain)-1)):
                self.grid[x][y] -= 1
            else:
                self.grid[x][y] -= 2

        self.full_chains[gen].append(chain)



    def perform_bell_measurements(self):
        # goes through each of the full chains and creates long range bell pairs
        for _, gen in enumerate(self.generators):
            for _, chain in enumerate(self.full_chains[gen.key]):
                for _, pair in enumerate(chain[1:len(chain)-1]):
                    x, y = pair
                    self.grid[x][y] += 2
                self.bell_pairs[gen.key].append([chain[0], chain[-1]])

            self.full_chains[gen.key] = []


    def perform_syndrome_measurements(self):
        # performs syndrome measurement on any fully routed generators and frees their bell pairs (and extra readout ancilla)
        for _, gen in enumerate(self.generators):
            if (not gen.qbts_to_route):
                # free readout qubit
                # x, y = gen.dest
                # self.grid[x][y] += 1

                # free bell pairs
                for _, pair in enumerate(self.bell_pairs[gen.key]):
                    for _, qbt in enumerate(pair):
                        x, y = qbt
                        self.grid[x][y] += 1

                self.bell_pairs[gen.key] = []


    @staticmethod
    def tmp_add_chain(grid, chain):
        # performs add chain on a provided ancilla grid
        for i, pair in enumerate(chain):
            x, y = pair
            if ((i == 0) or (i == len(chain)-1)):
                grid[x][y] -= 1
            else:
                grid[x][y] -= 2


    def print_grid(self):
        for i in range(self.M-1, -1, -1):
            for j in range(self.N):
                print(self.grid[i][j], end='')
            print()
        print()

    def set_available_ancillas(self, locs):
        for loc in locs:
            x,y = loc[0]
            if loc[1] > self.k: self.grid[x][y] = loc[1]

    @staticmethod
    def tmp_print_grid(grid):
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                print(grid[i][j], end='')
            print()
        print()

    def find_optimal_site():
        # geometric median
        pass

    def route_generator(self, gen, prior_dest=None):
        # attempts to route a generator, if so returns chains and dest
        # if not returns most routings it was able to do and dest
        # able to provide pre-existing destination site, could be in generator or not...

        out = []
        possible_dests = gen
        if (prior_dest):
            possible_dests = [prior_dest]

        for _, dest in enumerate(possible_dests):
            if ((not prior_dest) and self.in_progress[dest[0]][dest[1]]):
                continue

            tmp_grid = copy.deepcopy(self.grid)
            chains = []
            routed_qbts = []
            tot_len = 0

            # dest is the meeting site
            # if (not prior_dest):
            #     tmp_grid[dest[0]][dest[1]] -= 1

            for _, site in enumerate(gen):
                if ((dest != site) and (not self.dests[site[0]][site[1]])):
                    chain = self.find_chain(tmp_grid, dest, site)
                    if chain:
                        self.tmp_add_chain(tmp_grid, chain)
                        chains.append(chain)
                        routed_qbts.append(site)
                        tot_len += len(chain)
                    else:
                        continue
            out.append({"dest":dest, "chains":chains, "routed_qbts":routed_qbts, "num":len(routed_qbts), "len":tot_len})

        if (not prior_dest):
            out.sort(key=lambda x: (x["num"], -x["len"]), reverse=True)

        # print(out)
        # for i in out:
        #     for j in i["chains"]:
        #         print(j)
        #     print(".........")
        # print("+++++++++")
        return out[0] if out else out


    def greedy_route_set(self, gens):
        # given a set of generators, return the number of rounds it takes to fully route it
        # round is considered routing whatever you can, performing bell measurements, and maybe doing a syndrome meas
        # gens: array containing sets of points that make up a generator measurement
        # prioritize finishing a single generator

        rounds = 0
        # self.reset(self.N, self.k)
        # gens = sorted(gens, key=lambda x: len(x[0]), reverse=False)

        for i, qbts in enumerate(gens):
            # if (len(qbts[0]) > self.k):
            #     print("Impossible to route") # possible if meeting at site in gen, if at external site impossible...
            #     return

            self.full_chains[i] = []
            self.bell_pairs[i] = []
            self.generators.append(Generator(qbts[0], qbts[1], i))

        # random.shuffle(self.generators)
        for _ in range(len(gens) + 1):
            # attempt to route closest to being done first
            # print(f"Round {_+1}: +++++++++++++++++++++++++++++ ")
            self.generators.sort(key=lambda x: len(x.routed), reverse=True)

            for gen in self.generators:
                res = self.route_generator(gen.qbts_to_route, gen.dest)
                print(res)

                if (res and res["chains"]):
                    if (not gen.dest):
                        dest = res["dest"]
                        # self.grid[dest[0]][dest[1]] -= 1
                        self.dests[dest[0]][dest[1]] = True
                        gen.dest = dest
                        gen.qbts_to_route.remove(dest)
                        gen.routed.append(dest)

                        # for qbt in gen.qbts:
                        #     self.in_progress[qbt[0]][qbt[1]] += 1

                    for qbt in res["routed_qbts"]:
                        gen.qbts_to_route.remove(qbt)
                        gen.routed.append(qbt)

                    for chain in res["chains"]:
                        self.add_chain(chain, gen.key)

                self.print_grid()
            # print(sum([sum(row) for row in self.grid]) / (self.N**2 * (self.k + 1)))

            self.perform_bell_measurements()
            self.perform_syndrome_measurements()

            for gen in self.generators:
                if (not gen.qbts_to_route):
                    self.dests[gen.dest[0]][gen.dest[1]] = False
                    for qbt in gen.qbts:
                        self.in_progress[qbt[0]][qbt[1]] -= 1
            self.generators[:] = [gen for gen in self.generators if gen.qbts_to_route]
            rounds += 1

            if not self.generators:
                return rounds
        return 0


    # def divided_greedy_route_set(self, gens, n):
    #     def chunk(it, size):
    #         it = iter(it)
    #         return iter(lambda: tuple(islice(it, size)), ())

    #     tot_rounds = 0
    #     for gen_list in list(chunk(gens, n)):
    #         rounds = self.greedy_route_set(gen_list)
    #         if rounds:
    #             tot_rounds += rounds
    #         else:
    #             return 0
    #     return tot_rounds

